In [2]:
import numpy as np
import pandas as pd
import random
from sklearn.model_selection import train_test_split

## Preprocessing Drug, Target and Drug-Target Interaction Matrics

In [3]:
drug_features = pd.read_csv("C:/Users/szest/Documents/DTI_prediction/Drug_features.txt",
                            delimiter="\t", header=None)
drug_ID = pd.read_csv("C:/Users/szest/Documents/DTI_prediction/Drug_ID.txt",
                        delimiter="\t", header=None)

drug_features_list = drug_features[0].values
drug_ID_list = drug_ID[0].values

drug = pd.read_csv("C:/Users/szest/Documents/DTI_prediction/Drug_feature_vectors.txt",
                        delimiter="\t", header=None)

drug.columns = drug_features_list
drug.index = drug_ID_list


In [4]:
target_features = pd.read_csv("C:/Users/szest/Documents/DTI_prediction/Target_features.txt",
                            delimiter="\t", header=None)
target_ID = pd.read_csv("C:/Users/szest/Documents/DTI_prediction/Target_ID.txt",
                        delimiter="\t", header=None)

target_features_list = target_features[0].values
target_ID_list = target_ID[0].values

target = pd.read_csv("C:/Users/szest/Documents/DTI_prediction/Target_feature_vectors.txt",
                        delimiter="\t", header=None)

target.columns = target_features_list
target.index = target_ID_list


In [5]:
interactions = pd.read_csv("C:/Users/szest/Documents/DTI_prediction/DTI.txt",
                        delimiter="\t", header=None)
interactions.columns = target_ID_list
interactions.index = drug_ID_list

In [6]:
drug

ALogP        ALogp2       AMR      apol        nA       nR   nN  \
DB00014  0.74017  1.967000e-02  0.248390  0.234090  0.230770  0.00000  0.0   
DB00035  0.77503  9.953400e-03  0.203240  0.188980  0.179490  0.16667  0.5   
DB00050  0.77162  1.075800e-02  0.291280  0.266010  0.256410  0.16667  0.0   
DB00091  0.84958  1.733700e-04  0.258190  0.251410  0.153850  0.00000  0.0   
DB00093  0.76321  1.288000e-02  0.201090  0.187440  0.205130  0.00000  0.5   
...          ...           ...       ...       ...       ...      ...  ...   
DB09202  0.86008  9.540100e-07  0.051852  0.042921  0.000000  0.00000  0.0   
DB09207  0.86391  1.205600e-05  0.082635  0.068223  0.000000  0.00000  0.0   
DB09217  0.87126  1.443600e-04  0.071860  0.059786  0.000000  0.00000  0.0   
DB09242  0.85360  7.231100e-05  0.047090  0.037685  0.000000  0.00000  0.0   
DB09272  0.86343  8.535200e-06  0.134260  0.108100  0.025641  0.00000  0.0   

          nD       nC       nQ  ...        MW    WTPT.1   WTPT.2    WTPT.3  \
DB00014  0.0  0.00000  0.00000  ...  0.204500  0.214260  0.84683  0.168440   
DB00035  0.0  0.16667  0.14286  ...  0.172050  0.173430  0.84123  0.144040   
DB00050  0.0  0.00000  0.00000  ...  0.230620  0.240360  0.84725  0.164940   
DB00091  0.0  0.00000  0.00000  ...  0.193680  0.192940  0.78691  0.123110   
DB00093  0.0  0.33333  0.14286  ...  0.167350  0.168840  0.84250  0.134010   
...      ...      ...      ...  ...       ...       ...      ...       ...   
DB09202  0.0  0.00000  0.00000  ...  0.033898  0.037519  0.88366  0.017294   
DB09207  0.0  0.00000  0.00000  ...  0.063400  0.055422  0.84046  0.038655   
DB09217  0.0  0.00000  0.00000  ...  0.053345  0.052963  0.83732  0.031491   
DB09242  0.0  0.00000  0.00000  ...  0.037940  0.036794  0.85124  0.038858   
DB09272  0.0  0.00000  0.00000  ...  0.091140  0.098923  0.85828  0.051718   

           WTPT.4    WTPT.5         WPATH      WPOL    XLogP    Zagreb  
DB00014  0.068113  0.319130  1.385100e-08  0.153850  0.61239  0.194730  
DB00035  0.057354  0.239180  7.539200e-09  0.126290  0.62466  0.153060  
DB00050  0.066911  0.296570  1.915800e-08  0.176810  0.63210  0.214290  
DB00091  0.057586  0.208420  9.877500e-09  0.172220  0.72813  0.175170  
DB00093  0.052540  0.222690  6.846000e-09  0.125140  0.63235  0.148810  
...           ...       ...           ...       ...      ...       ...  
DB09202  0.005873  0.036328  1.230200e-10  0.021814  0.66877  0.035714  
DB09207  0.010205  0.075116  3.616400e-10  0.045924  0.66390  0.050170  
DB09217  0.025298  0.000000  3.134900e-10  0.041332  0.67058  0.054422  
DB09242  0.005308  0.091323  1.142900e-10  0.024110  0.65869  0.034014  
DB09272  0.023952  0.088316  1.656900e-09  0.080367  0.66875  0.092687  

[5877 rows x 193 columns]

In [7]:
target

[G1.1.1.10]  [G1.1.1.11]  [G1.1.1.12]  [G1.1.1.13]  [G1.1.1.14]  \
A0R559       0.1375       0.0500       0.1250       0.0250       0.0250   
A1A2B5       0.1375       0.0000       0.0875       0.0500       0.0500   
A1AFW1       0.1250       0.0000       0.0625       0.0250       0.0250   
A2QLK4       0.0750       0.0000       0.1500       0.0625       0.0000   
A3FMN7       0.0625       0.0375       0.0750       0.0750       0.0625   
...             ...          ...          ...          ...          ...   
Q9ZGH3       0.1375       0.0000       0.1000       0.0250       0.0375   
Q9ZGI2       0.1000       0.0125       0.0875       0.1250       0.0125   
Q9ZHG9       0.1250       0.0000       0.0375       0.0375       0.0625   
Q9ZIV5       0.1625       0.0000       0.0500       0.0875       0.0125   
Q9ZLT0       0.0875       0.0125       0.0625       0.0750       0.0750   

        [G1.1.1.15]  [G1.1.1.16]  [G1.1.1.17]  [G1.1.1.18]  [G1.1.1.19]  ...  \
A0R559       0.1500       0.0375       0.0375       0.0000       0.1250  ...   
A1A2B5       0.0875       0.0250       0.0375       0.0250       0.0875  ...   
A1AFW1       0.1250       0.0250       0.0500       0.0750       0.0875  ...   
A2QLK4       0.2000       0.0625       0.0000       0.0000       0.0000  ...   
A3FMN7       0.0500       0.0125       0.1000       0.0875       0.0875  ...   
...             ...          ...          ...          ...          ...  ...   
Q9ZGH3       0.1125       0.0250       0.0500       0.0000       0.1375  ...   
Q9ZGI2       0.0625       0.0000       0.0375       0.0375       0.0750  ...   
Q9ZHG9       0.0750       0.0250       0.0750       0.0125       0.1000  ...   
Q9ZIV5       0.0625       0.0375       0.1125       0.0375       0.0500  ...   
Q9ZLT0       0.0875       0.0125       0.0750       0.0750       0.1250  ...   

        [G7.1.1.77]  [G7.1.1.78]  [G7.1.1.79]  [G7.1.1.7]  [G7.1.1.80]  \
A0R559      0.66218      0.49475      0.48147     0.60707      0.57762   
A1A2B5      0.51042      0.43465      0.48638     0.43520      0.50230   
A1AFW1      0.59517      0.68371      0.71606     0.34655      0.44717   
A2QLK4      0.46594      0.33616      0.52610     0.51987      0.53481   
A3FMN7      0.50774      0.92986      0.65366     0.39101      0.34645   
...             ...          ...          ...         ...          ...   
Q9ZGH3      0.69375      0.47668      0.56558     0.52022      0.56425   
Q9ZGI2      0.42579      0.62270      0.64493     0.44850      0.39072   
Q9ZHG9      0.45215      0.51367      0.49303     0.55548      0.67524   
Q9ZIV5      0.22205      0.62201      0.54431     0.62711      0.46601   
Q9ZLT0      0.36412      0.40925      0.60448     0.36399      0.37242   

        [G7.1.1.8]  [G7.1.1.9]   [G9.1]   [G9.2]   [G9.3]  
A0R559     0.43417     0.53013  0.56064  0.50991  0.65702  
A1A2B5     0.48996     0.57072  0.60663  0.58982  0.58493  
A1AFW1     0.50995     0.59692  0.37423  0.59176  0.58153  
A2QLK4     0.63286     0.54009  0.80863  0.40144  0.12138  
A3FMN7     0.33316     0.36020  0.68694  0.63613  0.54405  
...            ...         ...      ...      ...      ...  
Q9ZGH3     0.46340     0.55276  0.45878  0.59271  0.65143  
Q9ZGI2     0.64735     0.53685  0.59001  0.47333  0.45970  
Q9ZHG9     0.46675     0.44291  0.59356  0.58672  0.62366  
Q9ZIV5     0.47206     0.30005  0.49910  0.51422  0.52903  
Q9ZLT0     0.34272     0.34191  0.49096  0.67417  0.64164  

[3348 rows x 1290 columns]

In [8]:
interactions

A0R559  A1A2B5  A1AFW1  A2QLK4  A3FMN7  A5W4F4  A7B6P0  A7J8L3  \
DB00014       0       0       0       0       0       0       0       0   
DB00035       0       0       0       0       0       0       0       0   
DB00050       0       0       0       0       0       0       0       0   
DB00091       0       0       0       0       0       0       0       0   
DB00093       0       0       0       0       0       0       0       0   
...         ...     ...     ...     ...     ...     ...     ...     ...   
DB09202       0       0       0       0       0       0       0       0   
DB09207       0       0       0       0       0       0       0       0   
DB09217       0       0       0       0       0       0       0       0   
DB09242       0       0       0       0       0       0       0       0   
DB09272       0       0       0       0       0       0       0       0   

         A8MPY1  A9CJ63  ...  Q9Z9H6  Q9ZAG3  Q9ZB17  Q9ZEU2  Q9ZF99  Q9ZGH3  \
DB00014       0       0  ...       0       0       0       0       0       0   
DB00035       0       0  ...       0       0       0       0       0       0   
DB00050       0       0  ...       0       0       0       0       0       0   
DB00091       0       0  ...       0       0       0       0       0       0   
DB00093       0       0  ...       0       0       0       0       0       0   
...         ...     ...  ...     ...     ...     ...     ...     ...     ...   
DB09202       0       0  ...       0       0       0       0       0       0   
DB09207       0       0  ...       0       0       0       0       0       0   
DB09217       0       0  ...       0       0       0       0       0       0   
DB09242       0       0  ...       0       0       0       0       0       0   
DB09272       0       0  ...       0       0       0       0       0       0   

         Q9ZGI2  Q9ZHG9  Q9ZIV5  Q9ZLT0  
DB00014       0       0       0       0  
DB00035       0       0       0       0  
DB00050       0       0       0       0  
DB00091       0       0       0       0  
DB00093       0       0       0       0  
...         ...     ...     ...     ...  
DB09202       0       0       0       0  
DB09207       0       0       0       0  
DB09217       0       0       0       0  
DB09242       0       0       0       0  
DB09272       0       0       0       0  

[5877 rows x 3348 columns]

All values have been normalized according to their column. There are 5844 drugs with 193 features, 3348 targets (proteins) with 1290 features and the Drug-Target Interaction matrix has drugs on the rows and targets on the columns with 1 indicating interaction. Although 0 doesn't necessarily mean that there is no interaction, a large proportion of interactions labeled no interaction will truly have no interaction, hence we will treat all 0 and no interaction. 

In [9]:
drug.shape,target.shape,interactions.shape

((5877, 193), (3348, 1290), (5877, 3348))

## Dimensionality Reduction

Dimensionality Reduction will be carried out on both drug and target matrices with no. components chosen to maintain at least 95% variance. Singular Value Decomposition is often used to solve the "curse of dimensionality". SVD is a projection method that projects the data onto a lower-dimensional space while also preserving the most important relationships in the original data by preserving the components of the highest variance. 

In [10]:
from sklearn.decomposition import TruncatedSVD
from scipy.sparse import csr_matrix

# Make sparse matrix
drug_sparse = csr_matrix(drug)
target_sparse = csr_matrix(target)

In [11]:
tsvd = TruncatedSVD(n_components=drug_sparse.shape[1]-1)
drug_tsvd = tsvd.fit(drug)
tsvd_2 = TruncatedSVD(n_components=target_sparse.shape[1]-1)
target_tsvd = tsvd_2.fit(target)
tsvd_var_ratios = tsvd.explained_variance_ratio_
tsvd_2_var_ratios = tsvd_2.explained_variance_ratio_
var_explained = tsvd.explained_variance_ratio_.sum()
var_2_explained = tsvd_2.explained_variance_ratio_.sum()

In [12]:
def select_n_components(var_ratio, goal_var: float) -> int:
    # Set initial variance explained so far
    total_variance = 0.0
    
    # Set initial number of features
    n_components = 0
    
    # For the explained variance of each feature:
    for explained_variance in var_ratio:
        
        # Add the explained variance to the total
        total_variance += explained_variance
        
        # Add one to the number of components
        n_components += 1
        
        # If we reach our goal level of explained variance
        if total_variance >= goal_var:
            # End the loop
            break
            
    # Return the number of components
    return n_components

In [13]:
select_n_components (tsvd_var_ratios,0.95)

56

In [14]:
select_n_components (tsvd_2_var_ratios,0.95)

222

In [15]:
tsvd_drug = TruncatedSVD(n_components=56)
drug_reduced = tsvd_drug.fit_transform(drug)

tsvd_target = TruncatedSVD(n_components=222)
target_reduced = tsvd_target.fit_transform(target)

print (tsvd_drug.explained_variance_ratio_.sum())
print (tsvd_target.explained_variance_ratio_.sum())


0.9520175860658526
0.9499224241076847


The reduced target and drug arrays are as below, with drug having 67 components and target having 273
components

In [16]:
print(len(target_reduced),len(target_reduced[1]))
print(len(drug_reduced),len(drug_reduced[1]))

3348 222
5877 56


## Create Training instances (drug-target)

In [17]:
interactions_arr = interactions.to_numpy()

In [18]:
target_feats, drug_feats = [], []
target_name, drug_name , outputs = [], [], []

for drug_idx, drug_ID in enumerate(drug_ID_list):
    for target_idx, target_ID in enumerate(target_ID_list):
        drug_name.append(drug_ID)
        target_name.append(target_ID)
        outputs.append(interactions_arr[drug_idx, target_idx])
        target_feats.append(target_reduced[target_idx])
        drug_feats.append(drug_reduced[drug_idx])

In [19]:
data = pd.DataFrame({'Drug' : drug_name, 'Target' : target_name, "Interactions" : outputs})
data

Drug  Target  Interactions
0         DB00014  A0R559             0
1         DB00014  A1A2B5             0
2         DB00014  A1AFW1             0
3         DB00014  A2QLK4             0
4         DB00014  A3FMN7             0
...           ...     ...           ...
19676191  DB09272  Q9ZGH3             0
19676192  DB09272  Q9ZGI2             0
19676193  DB09272  Q9ZHG9             0
19676194  DB09272  Q9ZIV5             0
19676195  DB09272  Q9ZLT0             0

[19676196 rows x 3 columns]

In [20]:
data_arr=data.to_numpy()

Because of the size of the training data, it was impractical to concatenate the data matrix with the corresponding reduced drug and target feature matrices.The target feature vector has 273 features, the drug feature vector has 67 features and are both created to match the drug-target interactions, hence both vectors have 196776196 samples. The drug and target name vectors have been incorporated into the data_arr, hence data_arr has 3 columns. 

In [21]:
print(len(target_feats),len(target_feats[1])) # has target features
print(len(drug_feats),len(drug_feats[1])) #has drug features
print(len(data_arr),len(data_arr[1])) #has drug name, target name and interaction

19676196 222
19676196 56
19676196 3


Sanity Check

In [22]:
data_arr[5000] #array(['DB00035', 'P35916', 0], dtype=object)
drug.loc[:"DB00035"] #second row
drug_reduced[1]==drug_feats[5000] 
target.loc[:"P35916"] #1653
target_reduced[1652]==target_feats[5000]

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,

## Addressing Between-Class bias

There are 12674 positive samples and 19663522 negative samples. The highly-imbalanced dataset will bias the model towards a negative interaction. In order to solve this, I will be segmenting the dataset into positive and negative datasets and using the same no of positive and negative samples to train the data.

In [23]:
#Positive and negative data on drug, target and interactions
positive = data[data["Interactions"]==1] 
positive_indexes = list(positive.index)
negative=data[data["Interactions"]==0]
negative_indexes=list(negative.index)
print(len(positive),len(negative))


12674 19663522


In [24]:
#Positive data on drug, target features 
drug_positive, target_positive = [], []
for i in positive_indexes:
    drug_positive.append(drug_feats[i])
    target_positive.append(target_feats[i])

In [25]:
# Comnbine positive data of drug and target features
positive_features= []
positive_features=np.concatenate((drug_positive,target_positive),axis=1)
print(len(positive_features),len(positive_features[0])) 

12674 278


In [26]:
#Keep track of drug and target pairs too as well as interaction
pos_ID_interaction = []
for i in positive_indexes:
    pos_ID_interaction.append(data_arr[i])

print(len(pos_ID_interaction),len(pos_ID_interaction[0]))

12674 3


There are too many negative_indexes, I will take approximatley 10000000 calculated by if I do a max of 500 estimators. I will randomly choose 600000 negative features. Random.choice is used for sampling without replacement.

In [27]:
np.random.seed(42)
final_neg_indexes = np.random.choice(negative_indexes,size=10000000,replace=False)
len(final_neg_indexes) == len(set(final_neg_indexes)) # No duplicates for sanity check

True

In [28]:
#Negative data on drug, target features
drug_negative, target_negative = [], []
for i in final_neg_indexes:
    drug_negative.append(drug_feats[i])
    target_negative.append(target_feats[i])

In [ ]:
#Combine negative data of drug and target features
negative_features=[]
negative_features=np.concatenate((drug_negative,target_negative),axis=1)
print(len(negative_features),len(negative_features[0]))

In [29]:
#Keep track of drug and target pairs too as well as interaction
neg_ID_interaction= []
for i in final_neg_indexes:
    neg_ID_interaction.append(data_arr[i])

print(len(neg_ID_interaction),len(neg_ID_interaction[0]))

10000000 3


In [43]:
#Keep track of the positive features in the test and train set
positive=np.concatenate((pos_ID_interaction,positive_features),axis=1)

In [50]:
# I will now split the positive data to train and test 
pos_train,pos_test,pos_y,pos_y_test= train_test_split(positive,np.ones(12674,dtype=int),
                                                      test_size=0.20,random_state=42)
print(len(pos_train),len(pos_train[0]))
print(len(pos_test),len(pos_test[0]))

10139 281
2535 281


In [55]:
pos_training=pd.DataFrame(pos_train)
pos_training.drop(pos_training.iloc[:,0:3],inplace=True,axis=1)
pos_training_arr=pos_training.to_numpy()
print(len(pos_training_arr),len(pos_training_arr[0]))

10139 278


In [93]:
pos_testing=pd.DataFrame(pos_test)
pos_testing.drop(pos_testing.iloc[:,0:3],inplace=True,axis=1)
pos_testing_arr=pos_testing.to_numpy()
print(len(pos_testing_arr),len(pos_testing_arr[0]))


2535 278


## Addressing Within-Class Bias 

Since the negative samples can contain a lot of noise, I will only be addressing within-class bias within positive samples. The negative samples can contain noise because some of these samples may be undiscovered drug-target pairs instead of a confirmed negative interaction. I will conduct K-Means Clustering, I used the "elbow" and silhouette methods to find the optimal number of cluster. I then used SMOTE, synthetic minority oversampling technique to oversample the minority classes by duplicating minority class examples to match the size of the majority class. These new examples can be synthesized from the existing examples and hence don't add new information to the model. There is a limitation that the smallest cluster needs to have more samples than the n_neighbors. Hence, this is also a limiting factor on the number of cluster for K-Means. The higher the no. clusters, the less samples there are in each group. Hence, this can also increase the computational demands.

In [58]:
from imblearn.over_sampling import SMOTE
from sklearn.cluster import KMeans
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.model_selection import cross_val_score
from collections import Counter
from matplotlib import pyplot
from sklearn.metrics import silhouette_score

In [60]:
kmeans=KMeans(n_clusters=50,random_state=0).fit(pos_training_arr)
labels=kmeans.labels_
len(labels)

10139

Summarize Class Distribution

In [61]:
counter=Counter(labels)
print(counter)

Counter({1: 881, 11: 799, 12: 579, 33: 540, 46: 485, 3: 481, 20: 466, 24: 409, 39: 390, 6: 389, 35: 346, 43: 326, 16: 290, 42: 276, 15: 229, 37: 217, 18: 216, 25: 198, 28: 183, 9: 177, 13: 159, 26: 153, 45: 123, 2: 112, 21: 111, 36: 110, 27: 110, 4: 109, 5: 106, 14: 103, 10: 93, 8: 90, 22: 90, 7: 75, 19: 70, 44: 69, 38: 65, 34: 58, 40: 57, 23: 56, 31: 49, 17: 48, 47: 43, 30: 40, 49: 36, 0: 31, 29: 31, 41: 28, 48: 19, 32: 18})


In [63]:
oversample = SMOTE()
pos_final, new_labels = oversample.fit_resample(pos_training_arr, labels)

In [64]:
print(len(pos_final),len(pos_final[0]))

44050 278


In [65]:
new_counter=Counter(new_labels)
print(new_counter)

Counter({36: 881, 46: 881, 11: 881, 21: 881, 42: 881, 26: 881, 12: 881, 27: 881, 25: 881, 1: 881, 20: 881, 9: 881, 16: 881, 39: 881, 45: 881, 37: 881, 3: 881, 8: 881, 6: 881, 35: 881, 33: 881, 43: 881, 24: 881, 18: 881, 44: 881, 7: 881, 22: 881, 19: 881, 14: 881, 13: 881, 4: 881, 48: 881, 28: 881, 23: 881, 40: 881, 10: 881, 15: 881, 5: 881, 34: 881, 41: 881, 47: 881, 49: 881, 0: 881, 2: 881, 31: 881, 32: 881, 29: 881, 38: 881, 17: 881, 30: 881})


In [67]:
no=len(pos_final)
no

44050

## Ensemble of Decision tree with Biases considered

In [68]:
#Combine negative data of drug and target features
temp=drug_negative[0:no]
temp2=target_negative[0:no]
neg_final=np.concatenate((temp,temp2),axis=1)
len(neg_final)

44050

In [71]:
X=np.concatenate((pos_final,neg_final),axis=0)
print(len(X),len(X[1]))

88100 278


In [70]:
pos=np.ones(no,dtype=int)
neg=np.zeros(no,dtype=int)
y=np.concatenate((neg,pos))
len(y)

88100

Boosting combines several weak learners into a strong leader and is used to reduce bias and variance. It does this through a weighted majority vote. Examples are Ada Boost,XGBoost and Gradient Tree Boost. Bootstrapping is sampling with replacement, leaving some data unselected so the unsampled training instances are Out of bag instances which can be used as validation. Bagging is bootstrap aggregating. Random forest and bagging classifiers. Once you have your sample from bootstrapping, a series of models can be build and this ensemble of medels will carry votes with equal weight. 

In order to increase diversity between the trees in order to have better performance, I will be using max_features and max_samples to tune the DecisionTreeClassifier. Boostrapping for samples and features will also be tested with gridsearch. 

Final Model: There is no bootstrapping- so features selected without replacement. There is feature subspacing and sample subspacing, with 0.2 and 0.5 of the original dataset selected for each tree. 

In [72]:
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import RepeatedKFold
seed=8
kfold = RepeatedKFold(random_state = seed)

model = BaggingClassifier(base_estimator = DecisionTreeClassifier(), 
                          n_estimators = 150, 
                          random_state = seed,max_features=0.2,max_samples=0.5,bootstrap=False,n_jobs=-2) 
results = cross_val_score(model, X, y, cv = kfold,scoring="roc_auc") 
print("AUC-ROC :") 
print(results.mean()) 

AUC-ROC :
0.9553297273487638


The AUC-ROC cross validation score is 0.95 after conducting gridsearch CV to tune the parameters: max_features, max_samples and bootstrap.

In [74]:
model.fit(X,y)

BaggingClassifier(base_estimator=DecisionTreeClassifier(), bootstrap=False,
                  max_features=0.2, max_samples=0.5, n_estimators=150,
                  n_jobs=-2, random_state=8)

Model is saved for testing and further predictions

In [130]:
import pickle
file_name="Decision Tree Ensemble.pkl"

with open(file_name,"wb") as file:
    pickle.dump(model,file)

## GridSearch CV

As mentioned before, GridSearchCV was conducted to tune bootstrap, max_samples and max_features. 

In [90]:
from sklearn.model_selection import RandomizedSearchCV
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
seed=8
param_grid = {
    'max_samples' : [0.4, 0.5],
    'max_features':[0.1,0.2],
}

chosen_scoring=("roc_auc")

clf = GridSearchCV(BaggingClassifier(DecisionTreeClassifier(),
                                     n_estimators = 150,  random_state = seed,n_jobs=-2,bootstrap=False,verbose=10),
                                       param_grid, scoring = chosen_scoring)
clf.fit(X, y)


[Parallel(n_jobs=11)]: Using backend LokyBackend with 11 concurrent workers.
[Parallel(n_jobs=11)]: Done   2 out of  11 | elapsed:   13.4s remaining:  1.0min
[Parallel(n_jobs=11)]: Done   4 out of  11 | elapsed:   14.0s remaining:   24.6s
[Parallel(n_jobs=11)]: Done   6 out of  11 | elapsed:   14.7s remaining:   12.2s
[Parallel(n_jobs=11)]: Done   8 out of  11 | elapsed:   14.9s remaining:    5.5s
[Parallel(n_jobs=11)]: Done  11 out of  11 | elapsed:   15.1s finished
[Parallel(n_jobs=11)]: Using backend LokyBackend with 11 concurrent workers.
[Parallel(n_jobs=11)]: Done   2 out of  11 | elapsed:    0.3s remaining:    1.6s
[Parallel(n_jobs=11)]: Done   4 out of  11 | elapsed:    0.4s remaining:    0.7s
[Parallel(n_jobs=11)]: Done   6 out of  11 | elapsed:    0.4s remaining:    0.3s
[Parallel(n_jobs=11)]: Done   8 out of  11 | elapsed:    0.4s remaining:    0.1s
[Parallel(n_jobs=11)]: Done  11 out of  11 | elapsed:    0.5s finished
[Parallel(n_jobs=11)]: Using backend LokyBackend with 11

[Parallel(n_jobs=11)]: Done   8 out of  11 | elapsed:    0.6s remaining:    0.2s
[Parallel(n_jobs=11)]: Done  11 out of  11 | elapsed:    0.7s finished
[Parallel(n_jobs=11)]: Using backend LokyBackend with 11 concurrent workers.
[Parallel(n_jobs=11)]: Done   2 out of  11 | elapsed:   24.6s remaining:  1.8min
[Parallel(n_jobs=11)]: Done   4 out of  11 | elapsed:   25.4s remaining:   44.5s
[Parallel(n_jobs=11)]: Done   6 out of  11 | elapsed:   25.5s remaining:   21.3s
[Parallel(n_jobs=11)]: Done   8 out of  11 | elapsed:   26.0s remaining:    9.7s
[Parallel(n_jobs=11)]: Done  11 out of  11 | elapsed:   26.5s finished
[Parallel(n_jobs=11)]: Using backend LokyBackend with 11 concurrent workers.
[Parallel(n_jobs=11)]: Done   2 out of  11 | elapsed:    0.3s remaining:    1.8s
[Parallel(n_jobs=11)]: Done   4 out of  11 | elapsed:    0.4s remaining:    0.8s
[Parallel(n_jobs=11)]: Done   6 out of  11 | elapsed:    0.5s remaining:    0.4s
[Parallel(n_jobs=11)]: Done   8 out of  11 | elapsed:   

[Parallel(n_jobs=11)]: Done   2 out of  11 | elapsed:    0.7s remaining:    3.4s
[Parallel(n_jobs=11)]: Done   4 out of  11 | elapsed:    0.8s remaining:    1.4s
[Parallel(n_jobs=11)]: Done   6 out of  11 | elapsed:    0.8s remaining:    0.7s
[Parallel(n_jobs=11)]: Done   8 out of  11 | elapsed:    0.9s remaining:    0.3s
[Parallel(n_jobs=11)]: Done  11 out of  11 | elapsed:    0.9s finished
[Parallel(n_jobs=11)]: Using backend LokyBackend with 11 concurrent workers.
[Parallel(n_jobs=11)]: Done   2 out of  11 | elapsed:   43.8s remaining:  3.3min
[Parallel(n_jobs=11)]: Done   4 out of  11 | elapsed:   46.5s remaining:  1.4min
[Parallel(n_jobs=11)]: Done   6 out of  11 | elapsed:   47.3s remaining:   39.4s
[Parallel(n_jobs=11)]: Done   8 out of  11 | elapsed:   47.7s remaining:   17.8s
[Parallel(n_jobs=11)]: Done  11 out of  11 | elapsed:   48.6s finished
[Parallel(n_jobs=11)]: Using backend LokyBackend with 11 concurrent workers.
[Parallel(n_jobs=11)]: Done   2 out of  11 | elapsed:   

GridSearchCV(estimator=BaggingClassifier(base_estimator=DecisionTreeClassifier(),
                                         bootstrap=False, n_estimators=150,
                                         n_jobs=-2, random_state=8,
                                         verbose=10),
             param_grid={'max_features': [0.1, 0.2], 'max_samples': [0.4, 0.5]},
             scoring='roc_auc')

In [94]:
clf.cv_results_


{'mean_fit_time': array([16.40621839, 26.23834982, 39.55158601, 52.88964205]),
 'std_fit_time': array([0.88798559, 3.33564003, 3.20729711, 7.27869796]),
 'mean_score_time': array([0.87100368, 0.95953293, 1.22401795, 1.26165929]),
 'std_score_time': array([0.05002436, 0.07332248, 0.05459476, 0.07869292]),
 'param_max_features': masked_array(data=[0.1, 0.1, 0.2, 0.2],
              mask=[False, False, False, False],
        fill_value='?',
             dtype=object),
 'param_max_samples': masked_array(data=[0.4, 0.5, 0.4, 0.5],
              mask=[False, False, False, False],
        fill_value='?',
             dtype=object),
 'params': [{'max_features': 0.1, 'max_samples': 0.4},
  {'max_features': 0.1, 'max_samples': 0.5},
  {'max_features': 0.2, 'max_samples': 0.4},
  {'max_features': 0.2, 'max_samples': 0.5}],
 'split0_test_score': array([0.80968053, 0.81178318, 0.8139563 , 0.82005817]),
 'split1_test_score': array([0.94759278, 0.95483813, 0.94500947, 0.95275561]),
 'split2_test_scor

In [91]:
clf.best_estimator_

BaggingClassifier(base_estimator=DecisionTreeClassifier(), bootstrap=False,
                  max_features=0.2, max_samples=0.5, n_estimators=150,
                  n_jobs=-2, random_state=8, verbose=10)

In [92]:
clf.best_params_

{'max_features': 0.2, 'max_samples': 0.5}

In [93]:
clf.best_score_

0.9397764550283341

## Initializing Testing

The test set is then created with the positive testing dataset concatenated with negative samples that werent included in training the model.

In [97]:
no_2=len(pos_testing_arr)
no_2

2535

In [98]:
temp=drug_negative[no+1:no+1+2535]
temp2=target_negative[no+1:no+1+2535]
neg_final_test=np.concatenate((temp,temp2),axis=1)
len(neg_final_test)

2535

In [100]:
X_test=np.concatenate((pos_testing_arr,neg_final_test),axis=0)
print(len(X_test),len(X_test[0]))

5070 278


In [105]:
pos=np.ones(no_2,dtype=int)
neg=np.zeros(no_2,dtype=int)
y_test=np.concatenate((neg,pos))
len(y_test)

5070

In [134]:
from numpy import savetxt
savetxt('X_test.csv',X_test,delimiter=",")
savetxt('y_test.csv',y_test,delimiter=",")


## Miscellaneous- Trying out Voting Classifier

In order to further increase diversity of the trees to increase the performance of the ensemble, different subset of the negative features can be used for each tree. Hence, this was tried below in tuning each individual tree. But because packages such as sklearn is so much more efficient, I reverted back to the methods above. 

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import VotingClassifier
from sklearn.metrics import accuracy_score

In [ ]:
n_estimators=100
n_val=27512
negative_arr=negative_features[0:n_estimators*n_val]
split_arr=np.array_split(negative_arr,n_estimators)


In [ ]:
model, scores, cv_scores, ind_scores = list(), list(), list(), list()

for i in range(n_estimators): 
    model.append((i,DecisionTreeClassifier(max_depth=None, max_features=0.2)))

for i in range(n_estimators):
    X_temp=np.concatenate((split_arr[i],X_res),axis=0)
    cv=RepeatedStratifiedKFold(n_splits=10,n_repeats=3,random_state=1)
    dt=model[i][1].fit(X_temp,y)
    #sc=cross_val_score(model[i][1],X_temp,y,cv=cv,n_jobs=-2)
    #cv_scores.append(sc)
    ind_scores.append(model[i][1].score(X_temp,y))
    scores.append(dt.predict(X_test))


In [ ]:
score_arr=np.array(scores)
most_voted=list()
for i in range(len(scores[0])):
    temp=score_arr[:,i]
    most_voted.append(np.bincount(temp).argmax())
    
acc=accuracy_score(y_test,most_voted)
acc